dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/EventAnalyticsController.java

```diff
@@ -393,22 +394,33 @@ public void getQueryHtmlCss(
     @GetMapping( value = RESOURCE_PATH + "/query/dimensions", produces = { APPLICATION_JSON_VALUE,
         "application/javascript" } )
     public AnalyticsDimensionsPagingWrapper<ObjectNode> getQueryDimensions(
-        @RequestParam String programStageId,
+        @RequestParam( required = false ) String programId,
+        @RequestParam( required = false ) String programStageId,
         @RequestParam( defaultValue = "*" ) List<String> fields,
         DimensionsCriteria dimensionsCriteria,
         HttpServletResponse response )
     {
+        validateRequest( programId, programStageId );
+
         configResponseForJson( response );
 
         List<PrefixedDimension> dimensions = eventAnalyticsDimensionsService
-            .getQueryDimensionsByProgramStageId( programStageId );
+            .getQueryDimensionsByProgramStageId( programId, programStageId );
 
         List<DimensionResponse> dimResponse = dimensionMapperService.toDimensionResponse( dimensions,
             EventAnalyticsPrefixStrategy.of( programStageId ) );
 
         return dimensionFilteringAndPagingService.pageAndFilter( dimResponse, dimensionsCriteria, fields );
     }
 
+    private void validateRequest( String programId, String programStageId )
+    {
+        if ( StringUtils.isBlank( programId ) && StringUtils.isBlank( programStageId ) )
+        {
+            throw new IllegalArgumentException( "Either programId or programStageId must be specified" );
```

Comment:
The exception message does not follow the logic.  I think both of them must not be blank.

Reply:
No, in this case the caller can pass program, program stage or both.
The invocation is unbelievable when both are blank